cache/src/main/java/io/envoyproxy/controlplane/cache/SnapshotResources.java

```diff
@@ -19,14 +22,38 @@
    */
   public static <T extends Message> SnapshotResources<T> create(Iterable<T> resources, String version) {
     return new AutoValue_SnapshotResources<>(
-        StreamSupport.stream(resources.spliterator(), false)
-            .collect(
-                Collector.of(
-                    ImmutableMap.Builder<String, T>::new,
-                    (b, e) -> b.put(Resources.getResourceName(e), e),
-                    (b1, b2) -> b1.putAll(b2.build()),
-                    ImmutableMap.Builder::build)),
-        version);
+        resourcesMap(resources),
+        version,
+        new HashMap<>());
+  }
+
+  /**
+   * Returns a new {@link SnapshotResources} instance with versions by resource name.
+   *
+   * @param resources the resources in this collection
+   * @param version the aggregated version for the resources in this collection
+   * @param versionsByResourceName map of versions by resource name
+   * @param <T> the type of resources in this collection
+   */
+  public static <T extends Message> SnapshotResources<T> create(
+      Iterable<T> resources,
+      String version,
```

Comment:
The map constructor is only used for EDS. Would it be better to create separate class for this? Let's say `EdsSnapshotResources` or more generic `SnapshotResourcesVersionedByName` (can't think of better name now). Then we would use it in `Snapshot` for EDS instead of `SnapshotResources`.

Reply:
Yeah I think we'd probably want a separate class, having unused parameters like this makes it a bit hard to understand what's going on. You'd probably want to add a parent interface to you can store either kind of resources, with appropriate `Snapshot::create` methods.